gms/gossiper.hh

```diff
@@ -248,22 +248,26 @@ private:
     // Must be called on shard 0.
     future<semaphore_units<>> lock_endpoint_update_semaphore();
 
+    struct live_and_unreachable_endpoints {
+        std::unordered_set<inet_address> live;
+        std::unordered_map<inet_address, clk::time_point> unreachable;
+    };
+
     // Must be called on shard 0.
     // Update _live_endpoints and/or _unreachable_endpoints
     // on shard 0, after acquiring `lock_endpoint_update_semaphore`.
-    // replicate_live_endpoints_on_change is called to replicate the changes to
-    // all shards and _live_endpoints_version is bumped unconditionally.
     //
-    // FIXME: The function is not exception safe so some shards
-    // may be left with stale _live_endpoints or _unreachable_endpoints.
-    // In this case, the _live_endpoints_version on that shard will not be updated.
-    // That can be examined under the endpoint_update_semaphore to make sure
-    // they all shards are consistent with each other.
-    future<> mutate_live_and_unreachable_endpoints(std::function<void(gossiper&)>);
+    // The called function returns a modified live_and_unreachable_endpoints
+    // that is then copied to temporary copies for all shards
+    // and then applied atomcically on all shards.
+    //
+    // The `on_success` callback is called on shard 0 iff replication was successful;
+    future<> mutate_live_and_unreachable_endpoints(std::function<live_and_unreachable_endpoints(gossiper&)> func,
```

Comment:
So: accept a l_a_u_e and return one (likely the same object, mutated),

Reply:
accepting a mutable reference and mutate it in place would be even simpler.

Reply:
done in v2